common/models/user.js

```diff
@@ -673,6 +673,49 @@ module.exports = function(User) {
       next();
     });
 
+    // Delete old sessions once email is updated
+    UserModel.observe('before save', function beforeEmailUpdate(ctx, next) {
+      if (ctx.isNewInstance) return next();
+      if (ctx.where) {
+        ctx.Model.find({ where: ctx.where }, function(err, userInstances) {
+          if (err) return next(err);
+          getOriginalUserData(userInstances);
+        });
+      } else if (ctx.instance) {
+        ctx.hookState.id = ctx.instance.id;
+        var oldId = ctx.hookState.id;
+        ctx.Model.find({ where: { id: oldId }}, function(err, userInstances) {
+          if (err) return next (err);
+          getOriginalUserData(userInstances);
+        });
+      }
+
+      function getOriginalUserData(userInstances) {
+        ctx.hookState.originalUserData = userInstances.map(function(u) {
+          return { id: u.id, email: u.email };
+        });
+      }
+      next();
+    });
+    UserModel.observe('after save', function afterEmailUpdate(ctx, next) {
+      if (!ctx.Model.relations.accessTokens) return next();
+      var AccessToken = ctx.Model.relations.accessTokens.modelTo;
+      var  newEmail = (ctx.instance || ctx.data).email;
+      if (!ctx.hookState.originalUserData) return next();
+      var idsToExpire = ctx.hookState.originalUserData.filter(function(u) {
+        return u.email !== newEmail;
```

Comment:
What if we are making a partial update and email is not changed at all?

```js
user.updateAttributes({ age: user.age+1 });
```

`newEmail` is `undefined`, `u.email` is the existing email, which means you will incorrectly invalidate all access tokens. What I find surprising is that you already have a test covering this path ([here](https://github.com/strongloop/loopback/pull/2693/files#diff-a202d1e6aac601eca4c15ad2418135fdR1894)). How comes it's passing? What am I missing?

Implementation wise, I think we should detect this situation early in "before save" hook to avoid unnecessary query.